cli/lib/kontena/cli/stacks/service_generator.rb

```diff
@@ -31,45 +31,43 @@ def parse_data(options)
-      data['shm_size'] = parse_memory(options['shm_size'].to_s) if options['shm_size']
-      data['cpus'] = options['cpus'] if options['cpus']
-      data['cpu_shares'] = options['cpu_shares'] if options['cpu_shares']
+      data['memory'] = options['mem_limit'] ? parse_memory(options['mem_limit'].to_s) : nil
+      data['memory_swap'] = options['memswap_limit'] ? parse_memory(options['memswap_limit'].to_s) : nil
+      data['shm_size'] = options['shm_size'] ? parse_memory(options['shm_size'].to_s) : nil
+      data['cpus'] = options['cpus'] ? options['cpus'] : nil
+      data['cpu_shares'] = options['cpu_shares'] ? options['cpu_shares'] : nil
       data['volumes'] = options['volumes'] || []
       data['volumes_from'] = options['volumes_from'] || []
-      data['cmd'] = Shellwords.split(options['command']) if options['command']
+      data['cmd'] = options['command'] ? Shellwords.split(options['command']) : []
       data['affinity'] = options['affinity'] || []
-      data['user'] = options['user'] if options['user']
+      data['user'] = options['user'] ? options['user'] : nil
       data['stateful'] = options['stateful'] == true
-      data['privileged'] = options['privileged'] unless options['privileged'].nil?
-      data['cap_add'] = options['cap_add'] if options['cap_add']
-      data['cap_drop'] = options['cap_drop'] if options['cap_drop']
-      data['net'] = options['net'] if options['net']
-      data['pid'] = options['pid'] if options['pid']
-      data['log_driver'] = options['log_driver'] if options['log_driver']
-      data['log_opts'] = options['log_opt'] if options['log_opt'] && !options['log_opt'].empty?
+      data['privileged'] = options['privileged'] || false
+      data['cap_add'] = options['cap_add'] ? options['cap_add'] : []
+      data['cap_drop'] = options['cap_drop'] ? options['cap_drop'] : []
+      data['net'] = options['net'] ? options['net'] : nil
+      data['pid'] = options['pid'] ? options['pid'] : nil
+      data['log_driver'] = options['log_driver'] ? options['log_driver'] : nil
+      data['log_opts'] = options['log_opt'] ? options['log_opt'] : {}
+      data['hooks'] = options['hooks'] || {}
+      data['secrets'] = options['secrets'] ? options['secrets'] : []
+      data['certificates'] = options['certificates'] ? options['certificates'] : []
+      data['build'] = options['build'] ? parse_build_options(options) : nil
+      data['health_check'] = parse_health_check(options)
+      data['stop_signal'] = options['stop_signal'] ? options['stop_signal'] : nil
+      data['stop_grace_period'] = options['stop_grace_period'] ? options['stop_grace_period'] : nil
+      data['read_only'] = options['read_only'] || false
+      data['entrypoint'] = options['entrypoint'] ? options['entrypoint'] : nil
+
       deploy_opts = options['deploy'] || {}
       data['strategy'] = deploy_opts['strategy'] if deploy_opts['strategy']
```

Comment:
The `strategy` is not reverting to the default it you remove it from the YAML... and changing this to send `{"strategy": null}` ends badly:

```
172.17.0.1 - - [07/Dec/2017:13:17:10 +0000] "PUT /v1/stacks/development/deploy-strategy HTTP/1.1" 200 1338 0.1151
I, [2017-12-07T13:17:11.025318 #17]  INFO -- GridServices::Update: updating service development/deploy-strategy/test with changes: strategy
172.17.0.1 - - [07/Dec/2017:13:17:11 +0000] "POST /v1/stacks/development/deploy-strategy/deploy HTTP/1.1" 200 153 0.1120
I, [2017-12-07T13:17:11.052678 #17]  INFO -- StackDeployWorker: removing following services: 
I, [2017-12-07T13:17:11.075384 #17]  INFO -- StackDeployWorker: deploying service development/deploy-strategy/test...
I, [2017-12-07T13:17:11.244300 #19]  INFO -- GridServiceSchedulerWorker: starting development/deploy-strategy/test deploy
E, [2017-12-07T13:17:11.253861 #19] ERROR -- : Actor crashed!
NoMethodError: undefined method `new' for nil:NilClass
	/app/app/jobs/grid_service_scheduler_worker.rb:89:in `deployer'
	/app/app/jobs/grid_service_scheduler_worker.rb:68:in `deploy'
	/app/app/jobs/grid_service_scheduler_worker.rb:13:in `block in watch'
	/app/app/jobs/grid_service_scheduler_worker.rb:11:in `loop'
	/app/app/jobs/grid_service_scheduler_worker.rb:11:in `watch'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
172.17.0.1 - - [07/Dec/2017:13:17:12 +0000] "GET /v1/stacks/development/deploy-strategy/deploys/5a293f577c3294001115b6db HTTP/1.1" 200 482 0.0988
172.17.0.1 - - [07/Dec/2017:13:17:12 +0000] "GET /v1/stacks/development/deploy-strategy/deploys/5a293f577c3294001115b6db HTTP/1.1" 200 482 0.0988
E, [2017-12-07T13:17:15.785039 #19] ERROR -- GridSchedulerJob: error occurred in service development/deploy-strategy/test
E, [2017-12-07T13:17:15.785152 #19] ERROR -- GridSchedulerJob: undefined method `new' for nil:NilClass
```